File: MergeSort.java
import java.io.*;
import java.util.*;

public class MergeSort {

    private static class DataElement {
        final int numericValue;
        final String textValue;

        public DataElement(int numericValue, String textValue) {
            this.numericValue = numericValue;
            this.textValue = textValue.trim();
        }
    }

    public static void main(String[] args) {
        if (args.length != 1) {
            System.err.println("Error: Please provide exactly one input file");
            System.out.println("Usage: java MergeSort <input_dataset.csv>");
            return;
        }

        String inputFilename = args[0];
        List<DataElement> dataset = loadDataset(inputFilename);

        long sortingStart = System.nanoTime();
        performMergeSort(dataset, 0, dataset.size() - 1);
        long sortingEnd = System.nanoTime();

        String outputFilename = "merge_sort_" + dataset.size() + ".csv";
        saveSortedResults(dataset, outputFilename);

        double runningTime = (sortingEnd - sortingStart) / 1_000_000_000.0;
        System.out.printf("Sorting completed time: " + runningTime + "seconds");
    }

    private static List<DataElement> loadDataset(String filename) {
        List<DataElement> elements = new ArrayList<>();

        try (BufferedReader fileReader = new BufferedReader(new FileReader(filename))) {
            String line;
            while ((line = fileReader.readLine()) != null) {
                String[] columns = line.split(",", 2);
                if (columns.length == 2) {
                    elements.add(new DataElement(
                        Integer.parseInt(columns[0].trim()),
                        columns[1]
                    ));
                }
            }
        } catch (IOException e) {
            System.err.println("Error processing file: " + e.getMessage());
            System.exit(1);
        }

        return elements;
    }

    private static void performMergeSort(List<DataElement> arr, int start, int end) {
        if (start < end) {
            int middle = (start + end) / 2;
            performMergeSort(arr, start, middle);
            performMergeSort(arr, middle + 1, end);
            mergeElements(arr, start, middle, end);
        }
    }

    private static void mergeElements(List<DataElement> arr, int start, int middle, int end) {
        List<DataElement> temporary = new ArrayList<>();
        int i = start, j = middle + 1;

        while (i <= middle && j <= end) {
            if (arr.get(i).numericValue <= arr.get(j).numericValue) {
                temporary.add(arr.get(i++));
            } else {
                temporary.add(arr.get(j++));
            }
        }

        while (i <= middle) temporary.add(arr.get(i++)); 
        while (j <= end) temporary.add(arr.get(j++));
        
        for (int k = 0; k < temporary.size(); k++) {
            arr.set(start + k, temporary.get(k));
        }
    }

    private static void saveSortedResults(List<DataElement> elements, String filename) {
        try (PrintWriter writer = new PrintWriter(new FileWriter(filename))) {
            for (DataElement element : elements) {
                writer.println(element.numericValue + "," + element.textValue);
            }
        } catch (IOException e) {
            System.err.println("Error writing output file: " + e.getMessage());
        }
    }
}

File: merge_sort.py
import sys
import time

class MergeSort:
    class DataElement:
        def __init__(self, numeric_value, text_value):
            self.numericValue = numeric_value
            self.textValue = text_value.strip()

    def __init__(self):
        pass

    def run(self, args):
        if len(args) != 1:
            print("Usage: python merge_sort.py dataset_n.csv")
            return

        input_filename = args[0]
        dataset = self.load_dataset(input_filename)

        sorting_start = time.time_ns()
        self.perform_merge_sort(dataset, 0, len(dataset) - 1)
        sorting_end = time.time_ns()

        output_filename = "merge_sort_" + str(len(dataset)) + ".csv"
        self.save_sorted_results(dataset, output_filename)

        running_time = (sorting_end - sorting_start) / 1_000_000_000.0
        print(f"Sorting completed time: {running_time}seconds")

    def load_dataset(self, filename):
        elements = []
        try:
            with open(filename, 'r') as file_reader:
                for line in file_reader:
                    columns = line.split(",", 1)
                    if len(columns) == 2:
                        elements.append(self.DataElement(
                            int(columns[0].strip()),
                            columns[1]
                        ))
        except IOError as e:
            print(f"Error processing file: {e}", file=sys.stderr)
            sys.exit(1)

        return elements

    def perform_merge_sort(self, arr, start, end):
        if start < end:
            middle = (start + end) // 2
            self.perform_merge_sort(arr, start, middle)
            self.perform_merge_sort(arr, middle + 1, end)
            self.merge_elements(arr, start, middle, end)

    def merge_elements(self, arr, start, middle, end):
        temporary = []
        i = start
        j = middle + 1

        while i <= middle and j <= end:
            if arr[i].numericValue <= arr[j].numericValue:
                temporary.append(arr[i])
                i += 1
            else:
                temporary.append(arr[j])
                j += 1

        while i <= middle:
            temporary.append(arr[i])
            i += 1

        while j <= end:
            temporary.append(arr[j])
            j += 1

        for k in range(len(temporary)):
            arr[start + k] = temporary[k]

    def save_sorted_results(self, elements, filename):
        try:
            with open(filename, 'w') as writer:
                for element in elements:
                    writer.write(f"{element.numericValue},{element.textValue}\n")
        except IOError as e:
            print(f"Error writing output file: {e}", file=sys.stderr)

if __name__ == "__main__":
    MergeSort().run(sys.argv[1:])
